src/storage/src/hummock/event_handler/cache_fill_policy.rs

```diff
@@ -0,0 +1,292 @@
+// limitations under the License.
+
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+use futures::future::try_join_all;
+use minstant::Instant;
+use risingwave_hummock_sdk::KeyComparator;
+use risingwave_pb::hummock::{group_delta, HummockVersionDelta};
+
+use crate::hummock::sstable_store::{SstableStoreRef, TableHolder};
+use crate::hummock::{Sstable, SstableBlockIterator};
+use crate::monitor::{CompactorMetrics, StoreLocalStatistic};
+
+pub struct CacheFillPolicy {
+    sstable_store: SstableStoreRef,
+    metrics: Arc<CompactorMetrics>,
+    cache_refill_io_count_limit: usize,
+    pending_preload: AtomicBool,
+}
+
+impl CacheFillPolicy {
+    pub fn new(
+        sstable_store: SstableStoreRef,
+        metrics: Arc<CompactorMetrics>,
+        cache_refill_io_count_limit: usize,
+    ) -> Self {
+        Self {
+            sstable_store,
+            metrics,
+            pending_preload: AtomicBool::new(false),
+            cache_refill_io_count_limit,
+        }
+    }
+
+    /// Only allow one thread call this method.
+    pub fn execute(self: &Arc<Self>, delta: HummockVersionDelta) {
+        if !self.pending_preload.load(Ordering::Acquire) {
+            self.pending_preload.store(true, Ordering::Release);
+            let policy = self.clone();
+            tokio::spawn(async move {
+                let start_time = Instant::now();
+                let _ = policy.fill_cache(delta).await;
+                policy.pending_preload.store(false, Ordering::Release);
+                policy
+                    .metrics
+                    .apply_version_duration
+                    .observe(start_time.elapsed().as_secs_f64());
+            });
+        }
+    }
+
+    async fn fill_cache(&self, delta: HummockVersionDelta) -> crate::hummock::HummockResult<()> {
+        let stats = StoreLocalStatistic::default();
+        let mut prefetch_blocks = vec![];
+        for group_delta in delta.group_deltas.values() {
+            let mut min_removed_level_idx = u32::MAX;
+            let mut flatten_reqs = vec![];
+            let mut removed_table_ids = vec![];
+            for d in &group_delta.group_deltas {
+                if let Some(group_delta::DeltaType::IntraLevel(level_delta)) = d.delta_type.as_ref()
+                {
+                    for sst in &level_delta.inserted_table_infos {
+                        flatten_reqs.push(self.sstable_store.sstable_syncable(sst, &stats));
+                    }
+                    if !level_delta.removed_table_ids.is_empty() {
+                        min_removed_level_idx =
+                            std::cmp::min(min_removed_level_idx, level_delta.level_idx);
+                    }
+                    removed_table_ids.extend(level_delta.removed_table_ids.clone());
+                }
+            }
+
+            if flatten_reqs.is_empty()
+                || min_removed_level_idx != 0
+                || self.cache_refill_io_count_limit == 0
```

Comment:
we can check `self.cache_refill_io_count_limit == 0` at the first of fill_cache / execute to avoid spawn task

Reply:
No. We must refill meta of sst even if we do not refill block cache